README.md
![](https://img.shields.io/github/workflow/status/avast/ep-stats/Code%20Checks?color=green)
[![PyPI version](https://img.shields.io/pypi/v/ep-stats?color=green)](https://pypi.org/project/ep-stats/)
[![Python versions](https://img.shields.io/pypi/pyversions/ep-stats?color=green)](https://pypi.org/project/ep-stats/)
[![Code style](https://img.shields.io/badge/formatted%20with-brunette-362511)](https://github.com/odwyersoftware/brunette)
[![Code style](https://img.shields.io/badge/styled%20with-flake8-green)](https://flake8.pycqa.org/en/latest/)
![](https://img.shields.io/github/languages/code-size/avast/ep-stats?color=green)
<img src="theme/experiment_b.png" align="right" />

# Ep-Stats

**Statistical package for experimentation platform.**

It provides a general python package and REST API that can be used to evaluate any metric
in AB test experiment.

## Features

* Robust two-tailed t-test implementation with multiple p-value corrections and delta-methods applied.
* Sequential evaluations allowing to stop experiments early.
* Connect it to any data source to either get pre-aggregated or per randomization unit data.
* Simple expression language to define arbitrary metrics.
* REST API to integrate it as a service in experimentation portal with score cards.

## Documentation

We have got a lovely [documentation](https://avast.github.io/ep-stats/).

## Base Example

Ep-stats allows for quick experiment evaluation. We are using provided testing data to evaluate metric `Click-through Rate` in experiment `test-conversion`.

```python
from epstats.toolkit import Experiment, Metric, SrmCheck
experiment = Experiment(
    'test-conversion',
    'a',
    [Metric(
        1,
        'Click-through Rate',
        'count(test_unit_type.unit.click)',
        'count(test_unit_type.global.exposure)'),
    ],
    [SrmCheck(1, 'SRM', 'count(test_unit_type.global.exposure)')],
    unit_type='test_unit_type')

# This gets testing data, use other Dao or get aggregated goals in some other way.
from epstats.toolkit.testing import TestData
goals = TestData.load_goals_agg(experiment.id)

# evaluate experiment
ev = experiment.evaluate_agg(goals)
```

`ev` contains evaluations of exposures, metrics and checks. This will have following output.

`ev.exposures`:

| exp_id | exp_variant_id | exposures |
| :----- | :------------- | --------: |
|test-conversion|a|21|
|test-conversion|b|26|

`ev.metrics`:

| exp_id | metric_id | metric_name | exp_variant_id | count | mean | std | sum_value | confidence_level | diff | test_stat | p_value | confidence_interval | standard_error | degrees_of_freedom |
| :----- | --------: | :---------- | -------------: | ----: | ---: | --: | --------: | ---------------: | ---: | --------: | ------: | ------------------: | -------------: | -----------------: |
|test-conversion|1|Click-through Rate|a|21|0.238095|0.436436|5|0.95|0|0|1|1.14329|0.565685|40|
|test-conversion|1|Click-through Rate|b|26|0.269231|0.452344|7|0.95|0.130769|0.223152|0.82446|1.18137|0.586008|43.5401|

`ev.checks`:

| exp_id | check_id | check_name | variable_id | value |
| :----- | -------: | :--------- | :---------- | ----: |
|test-conversion|1|SRM|p_value|0.465803|
|test-conversion|1|SRM|test_stat|0.531915|
|test-conversion|1|SRM|confidence_level|0.999000|

## Installation

You can install this package via `pip`.

```bash
pip install ep-stats
```

## Running

You can run a testing version of ep-stats via

```bash
python -m epstats
```

Then see Swagger on [http://localhost:8080/docs](http://localhost:8080/docs) for API documentation.

## Contributing

To get started locally, you can clone the repo and quickly get started using the `Makefile`.

```bash
git clone https://github.com/avast/ep-stats.git
cd ep-stats
make install-dev
```

It sets a new virtual environment `venv` in `./venv` using [venv](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/), installs all development dependencies, and sets [pre-commit](https://pre-commit.com/) git hooks to keep the code neatly formatted with [flake8](https://pypi.org/project/flake8/) and [brunette](https://pypi.org/project/brunette/).

To run tests, you can use `Makefile` as well.

```bash
source venv/bin/activate  # activate python environment
make check
```

To run a development version of ep-stats do

```bash
source venv/bin/activate
cd src
python -m epstats
```

### Documentation

To update documentation run

```bash
mkdocs gh-deploy
```

It updates documentation in GitHub pages stored in branch `gh-pages`.

## Inspiration

Software engineering practices of this package have been heavily inspired by marvelous [calmcode.io](https://calmcode.io/) site managed by [Vincent D. Warmerdam](https://github.com/koaning).


CHANGELOG.md
# Changelog #

## Version 1.0.1 ##
Features:
- Add CHANGELOG
- Final updates to README

Bug fixes:


src/epstats/server/res.py
import pandas as pd
from typing import List
from pydantic import BaseModel, Field

from ..toolkit import Evaluation
from .req import Experiment, Metric, Check


class MetricStat(BaseModel):
    """
    Per-variant metric evaluation result.
    """

    exp_variant_id: str = Field(title='Variant in the Experiment')
    diff: float = Field(
        title='Difference',
        description="""Relative difference of means of this variant and control variant.
        If this is a variant `b` and `a` is the control variant, then `diff = (b.mean - a.mean) / a.mean`.""",
    )
    mean: float = Field(
        title='Metric Mean',
        description="""Nominator and denominator to calculate the mean
        are given in metric definition. `mean = nominator / denominator`.""",
    )
    sum_value: float = Field(
        title='Metric Value',
        description="""Value of the metric, it is given by the
        nominator in the metric definition.""",
    )
    p_value: float = Field(
        title='p-Value',
        description="""We calculate p-value (under `confidence_level` statistical significance) of the relative
        difference (`diff`) of this variant mean and the control variant mean. We use
        [2-tailed Welch's test](https://en.wikipedia.org/wiki/Welch%27s_t-test)
        with unknown and unequal variance assumption and Welch–Satterthwaite equation approximation of degrees
        of freedom.""",
    )
    confidence_interval: float = Field(
        title='Confidence Interval',
        description="""Confidence interval for relative difference ('diff`)
        of means of this variant and control variant - `[mean - confidence_interval, mean + confidence_interval]`.
        Associated confidence level is the next parameter.""",
    )
    confidence_level: float = Field(
        title='Confidence Level (Statistical Significance)',
        description="""Confidence level used
        to compute (obtain) `confidence_interval`.""",
    )

    @staticmethod
    def from_df(df: pd.DataFrame):
        return [
            MetricStat(
                exp_variant_id=r['exp_variant_id'],
                diff=r['diff'],
                mean=r['mean'],
                sum_value=r['sum_value'],
                p_value=r['p_value'],
                confidence_interval=r['confidence_interval'],
                confidence_level=r['confidence_level'],
            )
            for i, r in df.iterrows()
        ]


class MetricResult(BaseModel):
    """
    Result of single metric evaluation.
    """

    id: int = Field(
        title='Metric Id', description="""Database id of the metric, not used at the moment in ep-stats""",
    )
    name: str = Field(
        title='Metric Name',
        description="""Official metric name as it appears in EP.
        The name is only for debugging and has no meaning for ep-stats.""",
    )
    stats: List[MetricStat] = Field(
        title='Per-variant statistics',
        description="""List with one entry per
        variant statistical results.""",
    )

    @staticmethod
    def from_df(metrics: List[Metric], df: pd.DataFrame):
        return [MetricResult(id=m.id, name=m.name, stats=MetricStat.from_df(df[df.metric_id == m.id])) for m in metrics]


class CheckStat(BaseModel):
    variable_id: str = Field(
        title='Check Variable',
        description="""Every check can return different
        variables and their values. E.g. SRM check returns `test_stat` and `p_value` variables with
        their `value`s.""",
    )
    value: float = Field(
        title='Value of the Variable',
        description="""Value of some variable returned by
        the check. E.g. SRM check returns `test_stat` and `p_value` variables with
        their `value`s.""",
    )

    @staticmethod
    def from_df(df: pd.DataFrame):
        return [CheckStat(variable_id=r['variable_id'], value=r['value']) for i, r in df.iterrows()]


class CheckResult(BaseModel):
    """
    Result of single check evaluation.
    """

    id: int = Field(
        title='Check Id', description='Database id of the check, not used at the moment.',
    )
    name: str = Field(
        title='Check Name',
        description="""Official check name as it appears in EP.
        The name is only for debugging and has no meaning for ep-stats.""",
    )
    stats: List[CheckStat] = Field(
        title='Per-variant statistics',
        description="""List with one entry per
        variant statistical results.""",
    )

    @staticmethod
    def from_df(checks: List[Check], df: pd.DataFrame):
        return [CheckResult(id=c.id, name=c.name, stats=CheckStat.from_df(df[df.check_id == c.id])) for c in checks]


class ExposureStat(BaseModel):
    """
    Exposures in the experiment per-variant.
    """

    exp_variant_id: str = Field(title='Variant in the Experiment')
    count: int = Field(
        title='Per-variant exposures', description="""Exposure count of experiment (randomization) unit.""",
    )

    @staticmethod
    def from_df(df: pd.DataFrame):
        return [ExposureStat(exp_variant_id=r['exp_variant_id'], count=r['exposures']) for i, r in df.iterrows()]


class ExposureResult(BaseModel):
    """
    Exposures in the experiment.
    """

    unit_type: str = Field(
        title='Experiment/randomization Unit Type',
        description="""Experiment (randomization) unit type is
    needed to correctly retrieve number of exposures per experiment variant.""",
    )
    stats: List[ExposureStat] = Field(
        title='Experiment Exposures', description="""List with experiment variant exposure counts per entry.""",
    )

    @staticmethod
    def from_df(experiment: Experiment, df: pd.DataFrame):
        return ExposureResult(unit_type=experiment.unit_type, stats=ExposureStat.from_df(df))


class Result(BaseModel):
    """
    Result of experiment evaluation.

    Top-level element in the response.
    """

    id: str = Field(title='Experiment Id',)
    metrics: List[MetricResult] = Field(
        title='Metric Results', description="""List with one entry per evaluated metric.""",
    )
    checks: List[CheckResult] = Field(
        title='Check Results', description="""List with one entry per evaluated check.""",
    )
    exposure: ExposureResult = Field(title='Experiment Exposures')

    @staticmethod
    def from_evaluation(experiment: Experiment, evaluation: Evaluation):
        metrics = MetricResult.from_df(experiment.metrics, evaluation.metrics)
        checks = CheckResult.from_df(experiment.checks, evaluation.checks)
        exposure = ExposureResult.from_df(experiment, evaluation.exposures)

        return Result(id=experiment.id, metrics=metrics, checks=checks, exposure=exposure)

    class Config:
        schema_extra = {
            'example': {
                'id': 'test-conversion',
                'metrics': [
                    {
                        'id': 1,
                        'name': 'Click-through Rate',
                        'stats': [
                            {
                                'exp_variant_id': 'a',
                                'diff': 0,
                                'mean': 0.23809523809523808,
                                'sum_value': 5,
                                'p_value': 1,
                                'confidence_interval': 1.1432928868841614,
                                'confidence_level': 0.95,
                            },
                            {
                                'exp_variant_id': 'b',
                                'diff': 0.13076923076923078,
                                'mean': 0.2692307692307692,
                                'sum_value': 7,
                                'p_value': 1,
                                'confidence_interval': 1.2327467657322932,
                                'confidence_level': 0.95,
                            },
                            {
                                'exp_variant_id': 'c',
                                'diff': 0.26,
                                'mean': 0.3,
                                'sum_value': 9,
                                'p_value': 1,
                                'confidence_interval': 1.352808784877644,
                                'confidence_level': 0.95,
                            },
                        ],
                    }
                ],
                'checks': [
                    {
                        'id': 1,
                        'name': 'SRM',
                        'stats': [
                            {'variable_id': 'p_value', 'value': 0.4528439055646014},
                            {'variable_id': 'test_stat', 'value': 1.5844155844155843},
                            {'variable_id': 'confidence_level', 'value': 0.999},
                        ],
                    }
                ],
                'exposure': {
                    'unit_type': 'test_unit_type',
                    'stats': [
                        {'exp_variant_id': 'a', 'count': 21},
                        {'exp_variant_id': 'b', 'count': 26},
                        {'exp_variant_id': 'c', 'count': 30},
                    ],
                },
            }
        }
